### ISSUE TITLE
When publishing a shiny app, you cannot use the Publish button in the IDE when the app dir has parentheses in the name

### ISSUE THREAD

**Issue by ajmcoqui** [rstudio/rsconnect#142]
If an application directory has parentheses in the name, and you attempt to deploy the app by clicking the Publish button in RStudio, you will get a 400 error.  If you remove the parentheses from the application directory name, the app will be deployed successfully.

STR
1. Open RStudio, File > New File > Shiny Web App
2. Give it a name without parentheses* and choose single-file app
3. Save the app, then rename the app dir with parentheses in the name
4. Open the app in RStudio, then click the Publish button
5. Point at a valid shinyapps.io account and publish
6. You will get an error (see attached)
7. Rename the app dir to something that has no parentheses
8. Open the app in RStudio, then click the Publish button.
9. Point at a valid shinyapps.io account and publish.  It will succeed.

* Note that if at this point you try to give it a name with parentheses in it, it won't let you create the app.
 
<img width="512" alt="screen shot 2016-11-17 at 4 19 03 pm" src="https://cloud.githubusercontent.com/assets/17852316/20407991/97a48a36-ace1-11e6-8aaa-fc9a271b2d06.png">

**Comment by ajmcoqui** [comment]
This is in the Deploy pane:
'Preparing to deploy application...Error: HTTP 400
POST https://api.shinyapps.io/v1/applications/
Validation Error
Execution halted'

Should I move this bug to the shinyapps.io project, or should rsconnect validate the app dir name before sending the app to shinyapps?

**Comment by jjallaire** [comment]
@jmcphers Where you do think this one belongs?

**Comment by jmcphers** [comment]
I think this belongs on the IDE. I'll take a look.

**Comment by kippandrew** [comment]
FYI: This is regex we use to validate application names on the server:  `^[A-Za-z0-9_-]{4,63}$`

**Comment by jmcphers** [comment]
Okay, the problem was indeed on the IDE side. We loosened up the validation because Connect supports a Title field which is almost entirely unvalidated, but in the process didn't cover all the cases where we needed to generate a stricter name for Shinyapps.io. I have now implemented destination-specific validation on the IDE side (i.e. you get on set of app name rules if you're publishing to shinyapps.io, and another set if going to Connect).
